### PR TITLE
CHE-9265: Implementation Theia terminal plugin.

### PR DESCRIPTION
**Pull request by AndrienkoAleksandr**
## Implementation Theia terminal plugin
This pull request implements server and client side for creation multi-machine terminals in the  Eclipse CHE workspaces(for docker environment). Client side and server side communicate with help websocket connections(protocol json-rpc). To manage all terminals used separated json-rpc connection, one per client. To send input/output of the terminal uses separated simple websocket connection.

### Terminal Server
Server side was written on the go-lang and uses docker cli to create terminal connections. Terminal connection based on docker exec.
Server side supported:

- Create new terminal exec;
- Connect to the created terminal exec;
- Resize terminal exec;
- Ability to reconnect to the created terminal exec with restore content. Restore content implemented like ring buffer with size 1000 lines.
Kill terminal doesn't supported yet, because docker api doesn't support kill exec. See more: [kill exec proposal](https://github.com/moby/moby/pull/9167)
For dependencies management used dep tool https://github.com/golang/dep.

### Terminal client
For now client side it's Theia Extension based on original Theia terminal extension, but modificated to support connection to the external server side. And this extension support ability select machine to terminal creation from command palette machine list. Minimal Theia core 0.3.8.

### CODE REVIEW

**Comment by AndrienkoAleksandr** [comment]
Don't review "vendor" directory it's external dependencies fixed by dep tool.

**Comment by AndrienkoAleksandr** [comment]
Needed https://github.com/eclipse/che/pull/9475

**Comment by benoitf** [comment]
> Don't review "vendor" directory it's external dependencies controlled by dep tool.
is this directory has to be checked in in the repository ? https://golang.github.io/dep/docs/FAQ.html#should-i-commit-my-vendor-directory

**Comment by AndrienkoAleksandr** [comment]
I think, we can save all used dependencies to have stable reproducible build protected with changes history of upstream projects. And for our go-lang agents we saved sources of external dependencies in the vendor folder, but we did it with help Godeps(for now this project is archived and recommends to use dep tool). We skips test tool and unused packages form upstream projects: 
    [prune]
      go-tests = true
      unused-packages = true

**Comment by eivantsov** [comment]
@skabashnyuk it won't.

**Comment by AndrienkoAleksandr** [comment]
CQ for docker(moby) cli https://dev.eclipse.org/ipzilla/show_bug.cgi?id=16187

File: che-theia-terminal/src/browser/terminal-widget/remote-terminal-widget.ts [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+            }
+
+            if (!size) {
+                return;
+            }
+
+            this.cols = size.cols;
+            this.rows = size.rows;
+
+            let resizeParam: ResizeParam = {
```
Comment (azatsarynnyy) [inline che-theia-terminal/src/browser/terminal-widget/remote-terminal-widget.ts]: A lot of variables within the PR are declared as `let` but they're actually constants. Should they be declared as `const`?

File: che-theia-terminal/src/browser/terminal-widget/remote-terminal-widget.ts [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+         * a refresh for example won't close it.  */
+        if (this.closeOnDispose === true && this.terminalId !== undefined) {
+            // do nothing, server side doesn't support kill exec.
+        }
+        super.dispose();
+    }
+
+    private async doResize() {
+        await Promise.all([this.waitForResized.promise, this.waitForTermOpened.promise]);
+        const geo = (this.term as any).proposeGeometry();
```
[the remark this answers is not in the document]
Comment (AndrienkoAleksandr) [reply che-theia-terminal/src/browser/terminal-widget/remote-terminal-widget.ts]: Nice catch.

File: che-theia-terminal/src/browser/terminal-widget/remote-terminal-widget.ts (hunk 1 above)
Comment (AndrienkoAleksandr) [reply che-theia-terminal/src/browser/terminal-widget/remote-terminal-widget.ts]: Yes, it should be.

**Comment by AndrienkoAleksandr** [comment]
CQ xterm.js 3.3.0 https://dev.eclipse.org/ipzilla/show_bug.cgi?id=16190 approved.

**Comment by AndrienkoAleksandr** [comment]
awaiting for CQ for docker(moby) cli.

**Pull request merged**